File: src/main/java/com/machfour/macros/cli/modes/Import.java
package com.machfour.macros.cli.modes;

import com.machfour.macros.cli.CommandImpl;
import com.machfour.macros.core.Schema;
import com.machfour.macros.core.datatype.TypeCastException;
import com.machfour.macros.objects.*;
import com.machfour.macros.storage.CsvImport;
import com.machfour.macros.storage.MacrosDataSource;

import java.io.FileReader;
import java.io.IOException;
import java.io.Reader;
import java.sql.SQLException;
import java.util.List;


import static com.machfour.macros.util.MiscUtils.toList;

public class Import extends CommandImpl {
    private static final String NAME = "import";
    private static final String USAGE = String.format("%s %s [--clear] [--norecipes] [--nofoods]", getProgramName(), NAME);

    public Import() {
        super(NAME, USAGE);
    }

    @Override
    public void printHelp() {
        out.println("Imports CSV data for foods, servings, recipes and ingredients into the database.");
        out.println("Only foods with index names not already in the database will be imported.");
        out.println("However, it will try to import all servings, and so will fail if duplicate servings exist.");
        out.println("Options:");
        out.println("  --clear       removes existing data before import");
        out.println("  --nofoods     prevents import of food and serving data (and clearing if --clear is used)");
        out.println("  --norecipes   prevents import of recipe data (and clearing if --clear is used)");

    }

    @Override
    public int doAction(List<String> args) {
        if (args.contains("--help")) {
            printHelp();
            return 0;
        }
        boolean doClear = args.contains("--clear");
        boolean noRecipes = args.contains("--norecipes");
        boolean noFoodsServings = args.contains("--nofoods");

        String foodCsvFile = config.getFoodCsvPath();
        String servingCsvFile = config.getServingCsvPath();
        String recipeCsvFile = config.getRecipeCsvPath();
        String ingredientsCsvFile = config.getIngredientsCsvPath();
        MacrosDataSource ds = config.getDataSourceInstance();

        try {
            if (doClear) {
                if (!noFoodsServings) {
                    out.println("Clearing existing foods, servings, nutrition data and ingredients...");
                    // have to clear in reverse order
                    ds.clearTable(Ingredient.table());
                    ds.clearTable(Serving.table());
                    ds.clearTable(NutritionData.table());
                    ds.clearTable(Food.table());
                } else if (!noRecipes) {
                    out.println("Clearing existing recipes and ingredients...");
                    // have to clear nutrition data first
                    ds.deleteByColumn(Food.table(), Schema.FoodTable.FOOD_TYPE, toList(FoodType.COMPOSITE.getName()));
                    ds.clearTable(Ingredient.table());
                } else {
                    out.println("Warning: nothing was cleared because both --nofoods and --norecipes were used");
                }
            }
            if (!noFoodsServings) {
                try (Reader foodCsv = new FileReader(foodCsvFile);
                     Reader servingCsv = new FileReader(servingCsvFile)) {
                    out.println("Importing foods and nutrition data into database...");
                    CsvImport.importFoodData(foodCsv, ds, false);
                    out.println("Saved foods and nutrition data");
                    CsvImport.importServings(servingCsv, ds, false);
                    out.println("Saved servings");
                    out.println();
                }
            }

            if (!noRecipes) {
                try (Reader recipeCsv = new FileReader(recipeCsvFile);
                     Reader ingredientsCsv = new FileReader(ingredientsCsvFile)) {
                    out.println("Importing recipes and ingredients into database...");
                    CsvImport.importRecipes(recipeCsv, ingredientsCsv, ds);
                    out.println("Saved recipes and ingredients");
                    out.println();
                }
            }
        } catch (SQLException e1) {
            out.println();
            err.println("SQL Exception occurred: " + e1.getMessage());
            return 1;
        } catch (IOException e2) {
            out.println();
            err.println("IO exception occurred: " + e2.getMessage());
            return 1;
        } catch (TypeCastException e3) {
            out.println();
            err.println("Type cast exception occurred: " + e3.getMessage());
            err.println("Please check the format of the CSV files");
            return 1;
        }

        out.println();
        out.println("Import completed successfully");
        return 0;
    }
}

File: src/main/java/com/machfour/macros/cli/modes/Import.kt
package com.machfour.macros.cli.modes

import com.machfour.macros.cli.CommandImpl
import com.machfour.macros.cli.utils.ArgParsingResult
import com.machfour.macros.cli.utils.findArgumentFromFlag
import com.machfour.macros.cli.utils.printlnErr
import com.machfour.macros.core.MacrosConfig
import com.machfour.macros.csv.CsvException
import com.machfour.macros.csv.importFoodData
import com.machfour.macros.csv.importRecipes
import com.machfour.macros.csv.importServings
import com.machfour.macros.entities.Food
import com.machfour.macros.entities.FoodNutrientValue
import com.machfour.macros.entities.Serving
import com.machfour.macros.queries.clearTable
import com.machfour.macros.queries.deleteAllCompositeFoods
import com.machfour.macros.queries.deleteAllIngredients
import com.machfour.macros.sql.datatype.TypeCastException
import java.io.FileReader
import java.io.IOException
import java.sql.SQLException

class Import(config: MacrosConfig) : CommandImpl(NAME, USAGE, config) {
    companion object {
        private const val NAME = "import"
        private const val USAGE = "Usage: $programName $NAME [--clear] [--norecipes] [--nofoods] " +
                " [-f <foods.csv>] [-s <servings.csv>] [-r <recipes.csv>] [-i <ingredients.csv>]"
    }

    override fun printHelp() {
        println("Imports CSV data for foods, servings, recipes and ingredients into the database.")
        println("Only foods with index names not already in the database will be imported.")
        println("However, it will try to import all servings, and so will fail if duplicate servings exist.")
        println()
        println("Options:")
        println("  --clear               removes existing data before import")
        println("  --nofoods             don't import food and serving data (if --clear is used, do not clear)")
        println("  --norecipes           don't import recipe data (if --clear is used, do not clear)")
        println("  -f <foods.csv>        read custom foods file (default: ${config.foodCsvPath}")
        println("  -s <servings.csv>     read custom servings file (default: ${config.servingCsvPath}")
        println("  -r <recipes.csv>      read custom recipes file (default: ${config.recipeCsvPath}")
        println("  -i <ingredients.csv>  read custom ingredients file (default: ${config.ingredientsCsvPath}")
        println()
        println("Note: If any custom CSV files are specified, no default paths will be used.")

    }


    override fun doAction(args: List<String>): Int {
        if (args.contains("--help")) {
            printHelp()
            return 0
        }
        val doClear = args.contains("--clear")
        val noRecipes = args.contains("--norecipes")
        val noFoodsServings = args.contains("--nofoods")
        val foodCsvArg = findArgumentFromFlag(args, "-f")
        val servingCsvArg = findArgumentFromFlag(args, "-s")
        val recipeCsvArg = findArgumentFromFlag(args, "-r")
        val ingredientCsvArg = findArgumentFromFlag(args, "-i")
        val foodCsvOverride: String?
        when (foodCsvArg) {
            is ArgParsingResult.KeyValFound -> foodCsvOverride = foodCsvArg.argument
            is ArgParsingResult.ValNotFound -> {
                printlnErr("Error - '-f' flag given but no food.csv specified")
                return 1
            }
            else -> {
                foodCsvOverride = null
            }
        }
        // TODO servings, etc
        println("servingCsvArg: $servingCsvArg")
        println("recipeCsvArg: $recipeCsvArg")
        println("ingredientCsvArg: $ingredientCsvArg")
        println()

        val foodCsvFile = foodCsvOverride ?: config.foodCsvPath
        val servingCsvFile = config.servingCsvPath
        val recipeCsvFile = config.recipeCsvPath
        val ingredientsCsvFile = config.ingredientsCsvPath
        val db = config.database
        try {
            if (doClear) {
                if (!noFoodsServings) {
                    println("Clearing existing foods, servings, nutrition data and ingredients...")
                    // have to clear in reverse order
                    // TODO Ingredients, servings, NutrientValues cleared by cascade?
                    deleteAllIngredients(db)
                    clearTable(db, Serving.table)
                    clearTable(db, FoodNutrientValue.table)
                    clearTable(db, Food.table)
                } else if (!noRecipes) {
                    println("Clearing existing recipes and ingredients...")
                    // nutrition data deleted by cascade
                    deleteAllIngredients(db)
                    deleteAllCompositeFoods(db)
                } else {
                    println("Warning: nothing was cleared because both --nofoods and --norecipes were used")
                }
            }
            if (!noFoodsServings) {
                println("Importing foods and nutrition data into database...")
                val conflictingFoods: Map<String, Food>
                FileReader(foodCsvFile).use {
                    conflictingFoods = importFoodData(db, it, false)
                }

                if (conflictingFoods.isNotEmpty()) {
                    println("The following foods will be imported; others had index names already present in the database:")
                    conflictingFoods.forEach { println(it.key) }
                }


                println("Saved foods and nutrition data")
                FileReader(servingCsvFile).use { importServings(db, it, false) }
                println("Saved servings")
                println()
            }
            if (!noRecipes) {
                println("Importing recipes and ingredients into database...")
                FileReader(recipeCsvFile).use { recipeCsv ->
                    FileReader(ingredientsCsvFile).use { ingredientsCsv ->
                        importRecipes(db, recipeCsv, ingredientsCsv)
                    }
                }
                println("Saved recipes and ingredients")
                println()
            }
        } catch (e1: SQLException) {
            println()
            printlnErr("SQL Exception occurred: ${e1.message}")
            return 1
        } catch (e2: IOException) {
            println()
            printlnErr("IO exception occurred: ${e2.message}")
            return 1
        } catch (e3: TypeCastException) {
            println()
            printlnErr("Type cast exception occurred: ${e3.message}")
            printlnErr("Please check the format of the CSV files")
            return 1
        } catch (e4: CsvException) {
            println()
            printlnErr("CSV import exception occurred: ${e4.message}")
            printlnErr("Please check the format of the CSV files")
        }

        println()
        println("Import completed successfully")
        return 0
    }
}
